File: ralph-workflow/src/guidelines/rust.rs
//! Rust-specific review guidelines
//!
//! Contains guidelines for Rust projects including web frameworks like Actix, Axum, and Rocket.

use super::base::ReviewGuidelines;
use crate::language_detector::ProjectStack;

/// Add Rust-specific guidelines to the review
pub(crate) fn add_guidelines(
    guidelines: ReviewGuidelines,
    stack: &ProjectStack,
) -> ReviewGuidelines {
    let base = ReviewGuidelines {
        quality_checks: guidelines
            .quality_checks
            .into_iter()
            .chain([
                "No unwrap/expect in production paths; use Result + ?".to_string(),
                "Proper lifetime annotations where needed".to_string(),
                "Prefer borrowing over cloning".to_string(),
                "Use strong types and exhaustive matching".to_string(),
                "Keep public API minimal (pub(crate) by default)".to_string(),
            ])
            .collect(),
        security_checks: guidelines
            .security_checks
            .into_iter()
            .chain([
                "Minimize unsafe code blocks; justify each use".to_string(),
                "Check for integer overflow in arithmetic".to_string(),
                "Validate untrusted input before processing".to_string(),
            ])
            .collect(),
        performance_checks: guidelines
            .performance_checks
            .into_iter()
            .chain([
                "Avoid unnecessary allocations (String → &str, Vec → slice)".to_string(),
                "Use iterators instead of indexing loops".to_string(),
                "Consider async for I/O-bound operations".to_string(),
            ])
            .collect(),
        testing_checks: guidelines
            .testing_checks
            .into_iter()
            .chain([
                "Unit tests for core logic (#[cfg(test)])".to_string(),
                "Integration tests in tests/ directory".to_string(),
                "Consider property-based testing for invariants".to_string(),
            ])
            .collect(),
        idioms: guidelines
            .idioms
            .into_iter()
            .chain([
                "Follow Rust API Guidelines".to_string(),
                "Use derive macros appropriately".to_string(),
                "Implement standard traits (Debug, Clone, etc.)".to_string(),
            ])
            .collect(),
        anti_patterns: guidelines
            .anti_patterns
            .into_iter()
            .chain([
                "Avoid .clone() to satisfy borrow checker without understanding".to_string(),
                "Don't use Rc<RefCell<T>> when ownership can be restructured".to_string(),
                "Avoid panic! in library code".to_string(),
            ])
            .collect(),
        ..guidelines
    };

    if stack
        .frameworks
        .iter()
        .any(|f| matches!(f.as_str(), "Actix" | "Axum" | "Rocket" | "Warp"))
    {
        add_rust_web_guidelines(base)
    } else {
        base
    }
}

fn add_rust_web_guidelines(guidelines: ReviewGuidelines) -> ReviewGuidelines {
    ReviewGuidelines {
        quality_checks: guidelines
            .quality_checks
            .into_iter()
            .chain([
                "Use extractors for request data".to_string(),
                "Handle errors with proper status codes".to_string(),
                "Use async handlers appropriately".to_string(),
            ])
            .collect(),
        security_checks: guidelines
            .security_checks
            .into_iter()
            .chain([
                "Validate all user input".to_string(),
                "Use tower middleware for common concerns".to_string(),
            ])
            .collect(),
        ..guidelines
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_rust_guidelines() {
        let stack = ProjectStack {
            primary_language: "Rust".to_string(),
            secondary_languages: vec![],
            frameworks: vec!["Actix".to_string()],
            has_tests: true,
            test_framework: Some("cargo test".to_string()),
            package_manager: Some("Cargo".to_string()),
        };

        let guidelines = add_guidelines(ReviewGuidelines::default(), &stack);

        // Should have Rust-specific checks
        assert!(guidelines
            .quality_checks
            .iter()
            .any(|c| c.contains("unwrap")));
        assert!(guidelines
            .security_checks
            .iter()
            .any(|c| c.contains("unsafe")));
        // Should have Actix-specific checks (web framework)
        assert!(guidelines
            .quality_checks
            .iter()
            .any(|c| c.contains("extractors")));
    }

    #[test]
    fn test_rust_without_web_framework() {
        let stack = ProjectStack {
            primary_language: "Rust".to_string(),
            secondary_languages: vec![],
            frameworks: vec!["Tokio".to_string()],
            has_tests: true,
            test_framework: Some("cargo test".to_string()),
            package_manager: Some("Cargo".to_string()),
        };

        let guidelines = add_guidelines(ReviewGuidelines::default(), &stack);

        // Should have Rust-specific checks but not web framework checks
        assert!(guidelines
            .quality_checks
            .iter()
            .any(|c| c.contains("unwrap")));
        // Should not have web-specific extractors check
        let extractor_count = guidelines
            .quality_checks
            .iter()
            .filter(|c| c.contains("extractors"))
            .count();
        assert_eq!(extractor_count, 0);
    }

    #[test]
    fn test_rust_axum_guidelines() {
        let stack = ProjectStack {
            primary_language: "Rust".to_string(),
            secondary_languages: vec![],
            frameworks: vec!["Axum".to_string()],
            has_tests: true,
            test_framework: Some("cargo test".to_string()),
            package_manager: Some("Cargo".to_string()),
        };

        let guidelines = add_guidelines(ReviewGuidelines::default(), &stack);

        // Should have Rust web framework checks
        assert!(guidelines
            .quality_checks
            .iter()
            .any(|c| c.contains("extractors") || c.contains("async")));
    }
}

File: ralph-python/ralph/guidelines/rust.py
"""Rust-specific review guideline categories.

Ported from the canonical Rust implementation in
``ralph-workflow/src/guidelines/rust.rs`` and adapted for the Python port.
The module exposes a lightweight data container that review prompt builders
can consume directly.
"""

from __future__ import annotations

from dataclasses import dataclass, field


@dataclass
class RustGuidelines:
    """Rust language-specific review checks.

    The categories mirror the review guideline structure used by the Rust
    implementation while including Rust-specific guidance around ownership,
    lifetimes, Clippy, panic-safety, and framework-oriented web handlers.
    """

    quality_checks: list[str] = field(
        default_factory=lambda: [
            "Code follows consistent style and formatting",
            "Functions have single responsibility",
            "Error handling is comprehensive",
            "No dead code or unused imports",
            "Run clippy and address actionable warnings",
            "No unwrap/expect in production paths; use Result + ?",
            "Proper lifetime annotations where needed",
            "Prefer borrowing over cloning",
            "Use strong types and exhaustive matching",
            "Keep public API minimal (pub(crate) by default)",
            "Use extractors for request data in web handlers",
            "Handle errors with proper status codes",
            "Use async handlers appropriately",
        ]
    )
    security_checks: list[str] = field(
        default_factory=lambda: [
            "No hardcoded secrets or credentials",
            "Input validation on external data",
            "Proper authentication/authorization checks",
            "Minimize unsafe code blocks; justify each use",
            "Check for integer overflow in arithmetic",
            "Validate untrusted input before processing",
            "Validate all user input",
            "Use tower middleware or equivalent for common web concerns",
        ]
    )
    performance_checks: list[str] = field(
        default_factory=lambda: [
            "No obvious performance bottlenecks",
            "Efficient data structures used",
            "Avoid unnecessary allocations (String → &str, Vec → slice)",
            "Use iterators instead of indexing loops",
            "Consider async for I/O-bound operations",
        ]
    )
    testing_checks: list[str] = field(
        default_factory=lambda: [
            "Tests cover main functionality",
            "Edge cases are tested",
            "Unit tests for core logic (#[cfg(test)])",
            "Integration tests in tests/ directory",
            "Consider property-based testing for invariants",
        ]
    )
    documentation_checks: list[str] = field(
        default_factory=lambda: [
            "Public APIs are documented",
            "Complex logic has explanatory comments",
            "Unsafe invariants and panic conditions are documented",
        ]
    )
    idioms: list[str] = field(
        default_factory=lambda: [
            "Code follows language conventions",
            "Follow Rust API Guidelines",
            "Use derive macros appropriately",
            "Implement standard traits (Debug, Clone, etc.) when justified",
        ]
    )
    anti_patterns: list[str] = field(
        default_factory=lambda: [
            "Avoid code duplication",
            "Avoid .clone() to satisfy the borrow checker without understanding ownership",
            "Do not reach for Rc<RefCell<T>> when ownership can be restructured",
            "Avoid panic! in library code",
        ]
    )
    concurrency_checks: list[str] = field(
        default_factory=lambda: [
            "Shared mutable state is properly synchronized",
            "No potential deadlocks from inconsistent lock ordering",
            "Spawned tasks respect Send/Sync requirements and cancellation paths",
        ]
    )
    resource_checks: list[str] = field(
        default_factory=lambda: [
            "Resources are properly closed or released on all paths",
            "No resource leaks in error paths",
            "Long-lived buffers, file handles, and sockets have clear ownership",
        ]
    )
    observability_checks: list[str] = field(
        default_factory=lambda: [
            "Errors are logged with context",
            "Critical operations have appropriate logging",
            "Structured logs do not omit error causes or request identifiers",
        ]
    )
    secrets_checks: list[str] = field(
        default_factory=lambda: [
            "Secrets loaded from environment/config, not hardcoded",
            "Sensitive data not logged or exposed in errors",
            "Credential material is zeroized or scoped tightly when feasible",
        ]
    )
    api_design_checks: list[str] = field(
        default_factory=lambda: [
            "API follows consistent naming conventions",
            "Breaking changes are clearly documented",
            "Return Result for recoverable failures instead of panicking",
            "Ownership, borrowing, and mutability are explicit in public APIs",
        ]
    )


__all__ = ["RustGuidelines"]
